### ISSUE TITLE
Input delay

### ISSUE THREAD

**Issue by paldepind** [mhartington/nvim-typescript#121]
# Problems summary

I get input lag when typing just before the completions appear. I don't experience this lag with other Deoplete sources.

## Expected

No noticeable input lag when typing.

## Environment Information

 * terminal: I can reproduce in both Konsole, neovim-qt and Vimr.
 * vim/nvim version: nvim v0.0.2

## Provide a minimal vim rc with less than 50 lines (Required!)

```vim
" Just to get Plug. Not relevant to the issue
if empty(glob('~/.config/nvim/autoload/plug.vim'))
silent !curl -fLo ~/.config/nvim/autoload/plug.vim --create-dirs
    \ https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
  autocmd VimEnter * PlugInstall
endif

call plug#begin()

function! DoRemote(arg)
  UpdateRemotePlugins
endfunction

" Actual configuration
Plug 'Shougo/deoplete.nvim', { 'do': function('DoRemote') } " Autocompletion
Plug 'HerringtonDarkholme/yats.vim'
Plug 'mhartington/nvim-typescript'

call plug#end()

let g:deoplete#enable_at_startup = 1
call deoplete#custom#source('typescript', 'min_pattern_length', 1)
let g:deoplete#auto_complete_delay = 0
```

## The reproduce ways from neovim starting

1. Open a TypeScript file.
2. Type three letters really fast. Like `xcv`.
3. There is a noticeable delay between the first and the second letter appearing. The `x` appears instantly and then after a very small delay `cv` appears at the same time. Note that I've set `min_pattern_length` to 1. I assume the delay would happen after a different character with a different setting.

Things to note:

* If I edit a TypeScript file with only Deopletes buffer completion there is zero delay.
* If I edit a Python file with `deoplete-jedi` there is zero delay.

## Example project repo?

I'm not sure if this is relevant. But I was testing in [this particular](https://github.com/funkia/list/blob/master/src/index.ts) TypeScript file.

**Comment by kreedz** [comment]
Same thing with `xcv`, and in insert mode characters removes not smoothly - with huge delay.

**Comment by mhartington** [comment]
Hmm, alright this seems to be due to calling vim functions in deoplete. This was done in order to only have one tsserver instance running, but I think i might have to revert that. I have something working, but would be able to test branch `fix-121`

**Comment by paldepind** [comment]
@mhartington Thanks a lot for taking a look at this @mhartington. The work you do on nvim-typescript is highly appreciated.

I've just tested the `fix-121` branch and I can confirm that the input lag I experienced seems to be gone :+1:

Only having one tsserver running is also pretty important (it can use a lot of memory). Hopefully some other solution can be found to that issue.

**Comment by mhartington** [comment]
Yeah, that was the goal with the original changes to the deoplete source. I had tried some other ideas using pythons multiprocess package, but had no luck. If you have any other ideas, I would love to hear them

**Comment by paldepind** [comment]
I'm afraid I can't be of much help. My Python is very basic. IMO it would be ideal if there was a good way to reuse the tsserver between plugins. I'm currently using both nvim-typescrit and ALE and it would be wonderful if they could share the same server. Maybe the ALE developer has ideas about how to make that happend?
